### DISCUSSION TITLE
请问 driver 和 adapter 是可以自由组合的么？

### DISCUSSION THREAD

**Discussion by luochen1990** [nonebot/discussions#487] [问答]
有此疑问是因为， 似乎每个 adapter 都会推荐一个不同的 driver 配置，我想知道的是，这里 adapter 的实现会跟 driver 相关么，用户在使用的时候，要根据使用的 adapter 不同而选择对应的 driver 么？   又或者其实相反，它们是正交的，可以自由组合？

**Comment by yanyongyu** [comment]
driver 之间可以有限组合，每种 driver 提供了某种功能。而 adapter 需要某种功能才可以工作，那他就需要具有某种功能的 driver。所以 adapter 实现只和提供某种功能的 driver 抽象相关，并不是直接依赖特定的 driver。

**Reply by luochen1990** [reply]
就是说不同 driver 实现了不同的 trait，然后不同的 adapter 依赖的 trait 也不一样，然后如果我用了某个 adapter，而它依赖了某个 trait， 我就得提供 实现了该 trait 的 driver ？   然后 driver 之间也会有冲突?  比如如果两个 driver 实现的同一个 trait 那么它们就不能共存？

这个逻辑是不是有点隐晦，我虽然能从抽象层面理解这个意思，但还是无法确认我需要哪些 driver， 特别是当同时使用多个 adapter 的情况下。。。

**Reply by luochen1990** [reply]
我想说的是，有没有可能，我们把这些 trait 整理出来，明确的标在 adapter 和 driver 的元信息里呢？

**Reply by yanyongyu** [reply]
并不存在冲突一说。驱动器类型这些信息均在文档有注明，请阅读文档驱动器部分。以及，如果你对Python主流框架熟悉，那你应该能从驱动器名称和描述上一眼看出他提供什么功能。

**Reply by luochen1990** [reply]
谢谢你的回复，但我需要解释一下

首先，我已经阅读过所有官方文档，特别是[这部分](https://nonebot.dev/docs/advanced/driver)，也正因为读了文档，我才发现驱动器有不同分类 ("驱动器类型大体上可以分为两种")，以及存在冲突的情况 （"服务端型驱动器仅能选择一个"）。

其次，既然该项目有意对 adapter 做抽象，那么是不是没必要要求用户对Python主流框架熟悉，从搞 nb-cli 项目我也能看出来降低用户门槛这个意图，正因此我才会在这里指出 adapter 这个抽象存在严重的 抽象泄漏 这一问题。

当然我也理解对 adapter 做抽象（即使它不太好抽象得很好且不存在抽象泄漏）可能是不希望把依赖写死，那么有没有可能利用 pip 的可选依赖 （或对等的 poetry 的 依赖组）之类的功能来做到这一点呢？

**Reply by yanyongyu** [reply]
对于你的第一个问题，server 类型驱动器只能选择一个这是符合逻辑的，并不存在“冲突”一说，如果你选择了两个 server 类型的驱动器，可以认为是人为错误。

用户可以不对框架熟悉，因此适配器给出了推荐驱动器配置，如果你对主流框架熟悉，那么你可以根据驱动器功能自由选择。

目前驱动器就是以可选依赖安装的，你可以通过 `nonebot2[driver-name]` 来安装对应驱动器。

**Reply by luochen1990** [reply]
那么我的更具体的问题则是，当你装了多个适配器，并且它们各自给出了不同的 推荐驱动器配置  的情况下，用户要怎么 合并这些 推荐驱动器配置 呢？

**Reply by yanyongyu** [reply]
你可以自由组合零个或一个 server 和多个 client 驱动器，配置格式为 `<module>[:<Driver>][+<module>[:<Mixin>]]*`。

**Reply by luochen1990** [reply]
对于上述问题，我可以根据文档猜测，合并过程大致上如下:

1. 在多份 “推荐驱动器配置” 里，把每份配置的 driver= 部分拆成 server-driver 和 client-driver 两个部分
2. 在多个 server-driver 里选择一个 （具体怎么选择我不清楚，也许任选一个都能工作？）
3. 将多分 client-driver 以集合的形式合并
4. 最后将合并过的 server-driver 和 client-driver 合并起来

这是一个没有被文档描述的隐晦的过程，我不知道这样干对不对。

主要是我认为，如果简单地把 server-driver 和 client-driver 的概念暴露给用户，事情将简单地多，且以上合并逻辑也将更加的不言自明。

**Reply by yanyongyu** [reply]
首先，你的合并过程是大致正确的，目前 server 只有 ASGI 类型所以任选即可，client 类型有重复功能合并 mixin 也不会影响使用。

关于你最后的建议，我认为已经在文档中说明清楚，参考 [驱动器类型](https://nonebot.dev/docs/advanced/driver#%E9%A9%B1%E5%8A%A8%E5%99%A8%E7%B1%BB%E5%9E%8B)。对于合并逻辑，在文档中也进行了说明并提供了示例，参考 [配置驱动器](https://nonebot.dev/docs/advanced/driver#%E9%85%8D%E7%BD%AE%E9%A9%B1%E5%8A%A8%E5%99%A8)。

我认为在阅读完此处文档后，应该对 server 和 client 类型 driver 有充分的 配置能力，并且在文档下方各驱动器介绍中，提供了该驱动器功能以及其类型的明确说明。

**Reply by luochen1990** [reply]
> adapter 实现只和提供某种功能的 driver 抽象相关，并不是直接依赖特定的 driver

对于这句话，我还有个追问是，这里的 “提供某种功能的 driver 抽象”  具体有哪些，是只有 client 和 server 的区别，还是说还有其他未在文档中提及的抽象（比如是否支持流式传输等等）?

**Reply by luochen1990** [reply]
> 我认为已经在文档中说明清楚，参考 [驱动器类型](https://nonebot.dev/docs/advanced/driver#%E9%A9%B1%E5%8A%A8%E5%99%A8%E7%B1%BB%E5%9E%8B)

文档中确实有说明，但用户是否能理解你的说明，并做出合理的推理，则不是必然的。

> 对于合并逻辑，在文档中也进行了说明并提供了示例，参考 [配置驱动器](https://nonebot.dev/docs/advanced/driver#%E9%85%8D%E7%BD%AE%E9%A9%B1%E5%8A%A8%E5%99%A8)

这份文档中确实有提到配置格式相关的部分，甚至这个格式也支持合并多个driver。 但并没有关于这些driver内在逻辑的描述，也没有对于我上面这段猜测的合并逻辑（或其等价/类似物）的描述

PS: 以上这些只是作为一个新用户，在使用时遇到了困扰，想给作者反馈一些值得改进的点。如果表达上不够礼貌请担待 :）

**Reply by yanyongyu** [reply]
具体有哪些你可以在[文档](https://nonebot.dev/docs/advanced/driver#%E9%A9%B1%E5%8A%A8%E5%99%A8%E7%B1%BB%E5%9E%8B)中看到，包括：

- ASGI 服务端驱动器
- HTTP 客户端驱动器
- WebSocket 客户端驱动器

其他类型目前没有提供。

**Reply by luochen1990** [reply]
> 关于你最后的建议

可能我没有明确地表述我的建议的措施是什么。 这里明确一下， 我的具体建议是:

1. 在配置里把 `driver=...` 替换为 `server-driver=...` 及 `client-drivers=...`
2. 在商店页面对 server driver 和 client driver 有明确的区分，比如提供醒目的标签，或能作为筛选条件

**Reply by yanyongyu** [reply]
在配置方面，nb 只需要 一个 server + 多个 client 形式即可，并没有什么复杂的合并逻辑，只需要将你需要的驱动器拼上就可以，nb 自己会处理 mixin。

**Reply by luochen1990** [reply]
> * ASGI 服务端驱动器
> * HTTP 客户端驱动器
> * WebSocket 客户端驱动器

谢谢，所以看起来这里分成3类，然后合并的时候其实每一类选择一个就可以了是么？  这么说我之前猜测的合并逻辑其实也不怎么正确（捂脸）

**Reply by yanyongyu** [reply]
关于区分问题，驱动器并不是非 server 即 client，因为还有啥都不是的驱动器（即 none 驱动器），driver 的概念是一体的，就是以零个或一个 server 带 多个 client 构成。

**Reply by yanyongyu** [reply]
你可以以 mixin 的角度来理解 driver，并不是每类只能选择一个，多选会覆盖而已。每驱动器就是实现了对应功能的 mixin。

**Reply by luochen1990** [reply]
有没有可能用 Protocol 来代替 mixin 建模呀？

**Reply by yanyongyu** [reply]
protocol 和 mixin 没有什么区别，只是在写法上不一样
